### PR TITLE
Improve error messages for raw strings (#60762)

### LINKED ISSUE #60762: error for too many trailing #s in raw string literals could be improved
for example ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=30cb101c03c070947da815e108333c93))
```rust
let foo = r##"bar"###;
```
results in
```
error: expected one of `.`, `;`, `?`, or an operator, found `#`
 --> src/main.rs:2:25
  |
2 |     let foo = r##"bar"###;
  |                         ^ expected one of `.`, `;`, `?`, or an operator here
```

while too few has a much nicer error:

```
error: unterminated raw string
 --> src/main.rs:3:15
  |
3 |     let baz = r##"quxx"#;
  |               ^ unterminated raw string
  |
  = note: this raw string should be terminated with `"##`
```

This issue has been assigned to @rcoh via [this comment](https://github.com/rust-lang/rust/issues/60762#issuecomment-605386317).

### PR DESCRIPTION
**Pull request by rcoh**
This diff improves error messages around raw strings in a few ways:
- Catch extra trailing `#` in the parser. This can't be handled in the lexer because we could be in a macro that actually expects another # (see test)
- Refactor & unify error handling in the lexer between ByteStrings and RawByteStrings
- Detect potentially intended terminators (longest sequence of "#*" is suggested)

Fixes #60762 
cc @estebank who reviewed the original (abandoned) PR for the same ticket.
r? @Centril

### CODE REVIEW

File: src/test/ui/parser/raw/raw_string.stderr [hunk 1]
```diff
@@ -2,7 +2,7 @@ error[E0748]: unterminated raw string
   --> $DIR/raw_string.rs:2:13
    |
 LL |     let x = r##"lol"#;
-   |             ^ unterminated raw string
+   |             ^       - help: you might have intended to terminate the string here: `##`
```
Comment (rcoh) [inline src/test/ui/parser/raw/raw_string.stderr]: not sure why the actual "message" got deleted but will figure it out.

File: src/librustc_lexer/src/cursor.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@ impl<'a> Cursor<'a> {
     /// If requested position doesn't exist, `EOF_CHAR` is returned.
     /// However, getting `EOF_CHAR` doesn't always mean actual end of file,
     /// it should be checked with `is_eof` method.
-    fn nth_char(&self, n: usize) -> char {
+    pub(crate) fn nth_char(&self, n: usize) -> char {
```
Comment (rcoh) [inline src/librustc_lexer/src/cursor.rs]: doesn't need to be public anymore, will remove.

File: src/librustc_lexer/src/lib.rs [hunk 3]
```diff
@@ -132,9 +136,65 @@ pub enum LiteralKind {
...
     ByteStr { terminated: bool },
     /// "r"abc"", "r#"abc"#", "r####"ab"###"c"####", "r#"a"
-    RawStr { n_hashes: usize, started: bool, terminated: bool },
+    RawStr(UnvalidatedRawStr),
     /// "br"abc"", "br#"abc"#", "br####"ab"###"c"####", "br#"a"
-    RawByteStr { n_hashes: usize, started: bool, terminated: bool },
+    RawByteStr(UnvalidatedRawStr),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
```
Comment (Centril) [inline src/librustc_lexer/src/lib.rs]: Consider adding doc comments here too.

File: src/librustc_lexer/src/lib.rs [hunk 4]
```diff
@@ -132,9 +136,65 @@ pub enum LiteralKind {
...
+pub struct UnvalidatedRawStr {
+    valid_start: bool,
+    n_start_hashes: usize,
+    n_end_hashes: usize,
+    possible_terminator_offset: Option<usize>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LexRawStrError {
+    /// Non # characters between `r` and `"` eg. `r#~"..`
```
Comment (Centril) [inline src/librustc_lexer/src/lib.rs]: Nit: for the rendering of this in rustdoc, add some more backticks here and in the fields below, e.g.,
```suggestion
    /// Non `#` characters between `r` and `"` eg. `r#~"..`
```

File: src/librustc_lexer/src/lib.rs [hunk 5]
```diff
@@ -132,9 +136,65 @@ pub enum LiteralKind {
...
+            (n_start, n_end) if n_start > n_end => Err(LexRawStrError::NoTerminator {
+                expected: n_start,
+                found: self.n_end_hashes,
+                possible_terminator_offset: self.possible_terminator_offset,
+            }),
+            (n_start, n_end) => {
+                debug_assert_eq!(n_start, n_end);
+                Ok(ValidatedRawStr { n_hashes: n_start_safe })
+            }
+        }
```
Comment (Centril) [inline src/librustc_lexer/src/lib.rs]: Some comments here could potentially be helpful.

File: src/librustc_lexer/src/lib.rs [hunk 6]
```diff
@@ -594,37 +656,49 @@ impl Cursor<'_> {
         false
     }
 
-    /// Eats the double-quoted string and returns a tuple of
-    /// (amount of the '#' symbols, raw string started, raw string terminated)
-    fn raw_double_quoted_string(&mut self) -> (usize, bool, bool) {
+    /// Eats the double-quoted string an UnvalidatedRawStr
```
Comment (Centril) [inline src/librustc_lexer/src/lib.rs]:
```suggestion
    /// Eats the double-quoted string an `UnvalidatedRawStr`.
```
Nit & general note: prefer backticks for type names and punctuate. :)

File: src/librustc_parse/lexer/mod.rs [hunk 7]
```diff
@@ -498,38 +510,52 @@ impl<'a> StringReader<'a> {
...
         }
 
+        if let Some(possible_offset) = possible_offset {
+            let span = self.mk_sp(
+                start + BytePos(possible_offset as u32),
+                start + BytePos(possible_offset as u32) + BytePos(found_terminators as u32),
+            );
+            err.span_suggestion(
+                span,
+                "you might have intended to terminate the string here",
```
Comment (Centril) [inline src/librustc_parse/lexer/mod.rs]:
```suggestion
                "consider terminating the string here",
```

File: src/librustc_parse/lexer/mod.rs [hunk 8]
```diff
@@ -498,38 +510,52 @@ impl<'a> StringReader<'a> {
...
-                "#".repeat(n_hashes as usize)
+                "#".repeat(n_hashes)
             ));
         }
 
+        if let Some(possible_offset) = possible_offset {
+            let span = self.mk_sp(
+                start + BytePos(possible_offset as u32),
+                start + BytePos(possible_offset as u32) + BytePos(found_terminators as u32),
+            );
```
Comment (Centril) [inline src/librustc_parse/lexer/mod.rs]:
```suggestion
            let lo = start + BytePos(possible_offset as u32);
            let span = self.mk_sp(lo, lo + BytePos(found_terminators as u32));
```

File: src/librustc_parse/parser/diagnostics.rs [hunk 9]
```diff
@@ -282,6 +286,31 @@ impl<'a> Parser<'a> {
         Err(err)
     }
 
+    fn check_too_many_raw_str_terminators(&mut self, err: &mut DiagnosticBuilder<'_>) -> bool {
+        let prev_token_raw_str = match self.prev_token {
+            Token { kind: TokenKind::Literal(Lit { kind: LitKind::StrRaw(n), .. }), .. } => Some(n),
+            Token {
+                kind: TokenKind::Literal(Lit { kind: LitKind::ByteStrRaw(n), .. }), ..
+            } => Some(n),
```
Comment (Centril) [inline src/librustc_parse/parser/diagnostics.rs]:
```suggestion
            Token { kind: TokenKind::Literal(Lit {
                kind: LitKind::StrRaw(n) | LitKind::ByteStrRaw(n),
                ..
             }), .. } => Some(n),
```
(the idea... rustfmt probably won't be happy with the formatting of the suggestion ^^)

File: src/librustc_parse/parser/diagnostics.rs [hunk 10]
```diff
@@ -282,6 +286,31 @@ impl<'a> Parser<'a> {
...
+            _ => None,
+        };
+
+        if let Some(n_hashes) = prev_token_raw_str {
+            if self.token.kind == TokenKind::Pound {
+                err.set_primary_message("too many `#` when terminating raw string");
+                err.span_suggestion(
+                    self.token.span,
+                    "Remove the extra `#`",
```
Comment (Centril) [inline src/librustc_parse/parser/diagnostics.rs]:
```suggestion
                    "remove the extra `#`",
```

File: src/librustc_parse/parser/diagnostics.rs [hunk 11]
```diff
@@ -282,6 +286,31 @@ impl<'a> Parser<'a> {
...
+                    String::new(),
+                    Applicability::MachineApplicable,
+                );
+                err.note(&format!("The raw string started with {} `#`s", n_hashes));
```
Comment (Centril) [inline src/librustc_parse/parser/diagnostics.rs]:
```suggestion
                err.note(&format!("the raw string started with {} `#`s", n_hashes));
```
Our convention is to start diagnostics with lowercase.

**Comment by Centril** [comment]
I've left some initial comments, but I'm not qualified to sign off on lexer changes myself.

r? @petrochenkov @matklad

File: src/librustc_parse/parser/diagnostics.rs [hunk 12]
```diff
@@ -282,6 +286,34 @@ impl<'a> Parser<'a> {
         Err(err)
     }
 
+    fn check_too_many_raw_str_terminators(&mut self, err: &mut DiagnosticBuilder<'_>) -> bool {
+        let prev_token_raw_str = match self.prev_token {
```
Comment (Centril) [inline src/librustc_parse/parser/diagnostics.rs]:
```suggestion
        let prev_token_raw_str = match self.prev_token.kind {
```
This should make the pattern matching a small bit nicer to read here.

Comment (rcoh) [reply src/librustc_parse/parser/diagnostics.rs]: smart.

File: src/librustc_lexer/src/lib.rs [hunk 13]
```diff
@@ -633,10 +722,23 @@ impl Cursor<'_> {
...
+            let n_end_hashes = self.eat_while(is_closing_hash);
+
+            if n_end_hashes == n_start_hashes {
+                return UnvalidatedRawStr {
+                    valid_start,
+                    n_start_hashes,
+                    n_end_hashes,
+                    possible_terminator_offset: None,
+                };
+            } else if n_end_hashes > 0 && n_end_hashes > max_hashes {
```
Comment (petrochenkov) [inline src/librustc_lexer/src/lib.rs]:
```suggestion
            } else if n_end_hashes > max_hashes {
```
Unnecessary condition.

File: src/librustc_parse/parser/diagnostics.rs [hunk 14]
```diff
@@ -282,6 +286,30 @@ impl<'a> Parser<'a> {
...
 
+    fn check_too_many_raw_str_terminators(&mut self, err: &mut DiagnosticBuilder<'_>) -> bool {
+        let prev_token_raw_str = match self.prev_token.kind {
+            TokenKind::Literal(Lit {
+                kind: LitKind::StrRaw(n) | LitKind::ByteStrRaw(n), ..
+            }) => Some(n),
+            _ => None,
+        };
+
+        if let Some(n_hashes) = prev_token_raw_str {
```
Comment (petrochenkov) [inline src/librustc_parse/parser/diagnostics.rs]:
```suggestion
        if let TokenKind::Literal(Lit { kind: LitKind::StrRaw(n_hashes) | LitKind::ByteStrRaw(n_hashes), .. }) = prev_token_raw_str {
```
Looks like the previous `match` can be merged into this `if let`.

Comment (rcoh) [reply src/librustc_parse/parser/diagnostics.rs]: ended up just merging it into one match

**Comment by petrochenkov** [comment]
@bors retry
(Looks like a network failure.)

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/70601#issuecomment-606579238, @bors r-

**Comment by CAD97** [comment]
cc @matklad is `librustc_lexer` really the right place to have this step for raw strings, or should that be up a level in `librustc_parser`'s lexer cooking? (My gut says the latter.)

**Comment by matklad** [comment]
This seem to fit pretty well into the rustc_lexer interface, which returns tokens as a bunch of sub-offsets. Ideally, we should dissolve lexer cooking altogether and instead lower the tokens much later, after macro expansion.

**Pull request merged**